yaml.go

```diff
@@ -307,24 +308,69 @@ type yamlError struct {
 	err error
 }
 
+func (e *yamlError) Unwrap() error {
+	return e.err
+}
+
 func fail(err error) {
-	panic(yamlError{err})
+	panic(&yamlError{err})
 }
 
 func failf(format string, args ...interface{}) {
-	panic(yamlError{fmt.Errorf("yaml: "+format, args...)})
+	panic(&yamlError{fmt.Errorf("yaml: "+format, args...)})
+}
+
+// ParserError is each error with a source line position found by the parser.
+// Unlike UnmarshalErrors, it is only used for terminal failures.
+type ParserError struct {
+	Message string
+	Line    int
+}
+
+func (e *ParserError) Error() string {
+	var b strings.Builder
+	b.WriteString("yaml: ")
+	if e.Line != 0 {
+		b.WriteString("line " + strconv.Itoa(e.Line) + ": ")
+	}
+	b.WriteString(e.Message)
+	return b.String()
+}
+
+// UnmarshalError is each error with a source position found by Unmarshal.
+type UnmarshalError struct {
+	Message string
```

Comment:
What do you guys think of replacing `Message string` with `Err error`? So any error values returned by `UnmarshalYAML` will be preserved as is.

Reply:
This is done. I've noticed that `gofmt` wants to do quite a bit more- it would make sende to `gofmt` everything in a separate PR before making any more changes.